Sched crash while creating node partition with indirect resource

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
scheduler crashes while creating node partition when the node resource it is looking at is an indirect resource.

#### Describe Your Change
Since in case of indirect resources, we do not set value of the resource on the resource structure. Scheduler should instead be using the resource structure of the node on which this indirect resource is dependent on.

#### Link to Design Doc
NA


#### Attach Test and Valgrind Logs/Output
[test_out.txt](https://github.com/PBSPro/pbspro/files/3283641/test_out.txt)



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bhroam commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,128 @@
+        Configure the PBS complex for node grouping test
+        """
+        # Create a non-consumable custom resource foostr'
+        self.server.add_resource(res, res_type, flag)
```
add_resource() is an old function that manually manipulates the resourcedef file.  Use self.server.manager() instead.

bhroam commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,128 @@
+        """
+
+        # Delete any existing vnodes on the system
+        self.server.manager(MGR_CMD_DELETE, NODE, None, "")
```
I think create_vnodes() does this for you.

bhroam commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,128 @@
+        # vnode[3], vnode[4] and vnode[5] respectively
+        attr1 = {'Resource_List.select': '1:foostr=A:vnode=vnode[3]'}
+        attr2 = {'Resource_List.select': '1:foostr=B:vnode=vnode[4]'}
+        attr3 = {'Resource_List.select': '1:foostr=C:vnode=vnode[5]'}
```
This isn't really testing node grouping.  It is testing indirect resources.  If you want to test node grouping, you can't request the node grouping resource.  Also requesting a specific vnode won't help either.  You need to submit 3 jobs requesting 2 nodes.  Make sure they're on the correct set of vnodes.

arungrover replied:
Thanks for this comment, It made me find another problem in the code where I was missing the change.

arungrover commented:
@bhroam as discussed offline, I've added an additional comment in the test explaining how nodes are grouped and what we are looking for. Please have a look.

arungrover commented:
Thanks @bhroam! I've changed the comment

agrawalravi90 commented on src/scheduler/node_partition.c:
```diff
@@ -423,6 +423,9 @@ create_node_partitions(status *policy, node_info **nodes, char **resnames, unsig
 				res = &unset_res;
 			}
 			if (res != NULL) {
+				/* Incase of indirect resource, point it to the right place */
```
im sure there's good reason to do it this way, i'm just curious, why don't we just make 'res' be 'indirect_res' when it is an indirect resource?

arungrover replied:
IMO it is because there will be multiple node objects pointing to the same object and when you free them you will have dangling pointers all over the place. Unless we decide to add a reference count to sched_resource structure.

agrawalravi90 commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,123 @@
+from tests.functional import *
+
+
+class TestSchedulingIndirectResources(TestFunctional):
```
instead of creating a specialized class for indirect resources, how about creating a more generic "TestIndirectResource" class?

agrawalravi90 commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,123 @@
+    Test Scheduler resolve indirect resources correctly
+    """
+
+    def config_complex_for_grouping(self, res, res_type='string', flag='h'):
```
you are calling this function just once, in 1 test, so i kind of feel that it should just be part of the test instead of being in a separate function.

arungrover replied:
In the future, When want to expand this test suite to add more tests for node grouping, these helper functions will be useful

agrawalravi90 commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,123 @@
+        """
+        # Create a custom resource
+        attr = {"type": res_type, "flag": flag}
+        self.server.manager(MGR_CMD_CREATE, RSC, attr, id=res)
```
there's also a server.add_resource() which basically does a qmgr, with some additional checking, i think. Just wondering if we should be using that interface to create the resource. What do you think?

arungrover replied:
It was server.add_resource() earlier but Bhroam said that that's an old interface and suggested to change it to this.

bhroam replied:
Oh, interesting, @agrawalravi90 is right.  server.add_resource() does use manager().  It existed before qmgr modified resources and modified the resourcedef file itself.  I thought it still did.  I guess it got rewritten at some point.  The only difference is that it keeps the PTL cache up to date.

agrawalravi90 commented on test/tests/functional/pbs_schedule_indirect_resources.py:
```diff
@@ -0,0 +1,123 @@
+        for i in range(3):
+            self.server.expect(JOB, {'job_state': 'R'}, id=j[i][0])
+            self.server.status(JOB, 'exec_vnode', j[i][0])
+            vn = j[i][1].get_vnodes()
```
this might be easier to read if you just added "jid, j = self.submit_job(attr)" in the for loop, and then referenced jid and j in the loop.

arungrover commented:
[test_fix.txt](https://github.com/PBSPro/pbspro/files/3297871/test_fix.txt)
[test_bug.txt](https://github.com/PBSPro/pbspro/files/3297872/test_bug.txt)
@agrawalravi90 I've addressed your comments, please have a look again. Also, attached are before and after test results.
